service/resource/deployment/resource.go

```diff
@@ -187,7 +190,85 @@ func (r *Resource) GetDeleteState(ctx context.Context, obj, currentState, desire
 }
 
 func (r *Resource) GetUpdateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
-	return nil, nil, nil, nil
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	currentDeployments, err := toDeployments(currentState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	desiredDeployments, err := toDeployments(desiredState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+
+	var deploymentsToCreate interface{}
+	{
+		deploymentsToCreate, err = r.GetCreateState(ctx, obj, currentState, desiredState)
+		if err != nil {
+			return nil, nil, nil, microerror.Mask(err)
+		}
+	}
+
+	var deploymentsToDelete []*v1beta1.Deployment
+	{
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "finding out which deployments have to be deleted")
+
+		for _, currentDeployment := range currentDeployments {
+			if !containsDeployment(desiredDeployments, currentDeployment) {
+				deploymentsToDelete = append(deploymentsToDelete, currentDeployment)
+			}
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", fmt.Sprintf("found %d deployments that have to be deleted", len(deploymentsToDelete)))
+	}
+
+	var deploymentsToUpdate []*v1beta1.Deployment
+	if updateallowedcontext.IsUpdateAllowed(ctx) {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "finding out which deployments have to be updated")
+
+		// Check if config maps of deployments changed. In case they did, add the
+		// deployments to the list of deployments intended to be updated.
+		m, ok := messagecontext.FromContext(ctx)
+		if ok {
+			for _, name := range m.ConfigMapNames {
+				desiredDeployment, err := getDeploymentByConfigMapName(desiredDeployments, name)
+				if err != nil {
+					return nil, nil, nil, microerror.Mask(err)
+				}
+				deploymentsToUpdate = append(deploymentsToUpdate, desiredDeployment)
+			}
+		}
+
+		// Check if deployments changed. In case they did, add the deployments to
+		// the list of deployments intended to be updated, but only in case they are
+		// not already being tracked.
+		for _, currentDeployment := range currentDeployments {
+			desiredDeployment, err := getDeploymentByName(desiredDeployments, currentDeployment.Name)
+			if IsNotFound(err) {
+				continue
+			} else if err != nil {
+				return nil, nil, nil, microerror.Mask(err)
+			}
+
+			if !isDeploymentModified(desiredDeployment, currentDeployment) {
+				continue
+			}
+
+			if containsDeployment(deploymentsToUpdate, desiredDeployment) {
+				continue
+			}
+
+			deploymentsToUpdate = append(deploymentsToUpdate, desiredDeployment)
+		}
```

Comment:
The tests proof that we add deployments to the update state that have changed itself. We also make sure that deployments are not tracked twice.